proxylib/proxylib/policymap.go

```diff
@@ -0,0 +1,332 @@
+		rule.AllowedRemotes[remote] = struct{}{}
+	}
+
+	// Each parser registers a parsing function to parse it's L7 rules
+	// This requires each L7 to use it's own oneof element!
+	l7Name := reflect.TypeOf(config.L7Rules).Elem().Name()
+	l7Parser, ok := L7RuleParsers[l7Name]
+	if ok {
+		log.Debugf("NPDS::PortNetworkPolicyRule: Calling L7Parser %s on %v", l7Name, config.String())
+		rule.L7Rules = l7Parser(config)
+	}
+	return rule, l7Name, ok
+}
+
+func (p *PortNetworkPolicyRule) Matches(remoteId uint32, l7 interface{}) bool {
+	// Remote ID must match if we have any.
+	if len(p.AllowedRemotes) > 0 {
+		_, found := p.AllowedRemotes[uint64(remoteId)]
+		if !found {
+			return false
+		}
+	}
+	if len(p.L7Rules) > 0 {
+		for _, rule := range p.L7Rules {
+			if rule.Matches(l7) {
+				log.Debugf("NPDS::PortNetworkPolicyRule: L7 rule matches (%v)", p)
+				return true
+			}
+		}
+		return false
+	}
+	// Empty set matches any payload
+	log.Debugf("NPDS::PortNetworkPolicyRule: Empty L7Rules matches (%v)", p)
+	return true
+}
+
+type PortNetworkPolicyRules struct {
+	Rules       []PortNetworkPolicyRule
+	HaveL7Rules bool
+}
+
+func newPortNetworkPolicyRules(config []*cilium.PortNetworkPolicyRule) (PortNetworkPolicyRules, bool) {
+	rules := PortNetworkPolicyRules{
+		Rules:       make([]PortNetworkPolicyRule, 0, len(config)),
+		HaveL7Rules: false,
+	}
+	if len(config) == 0 {
+		log.Debugf("NPDS::PortNetworkPolicyRules: No rules, will allow everything.")
+	}
+	var firstTypeName string
+	for idx, rule := range config {
+		newRule, typeName, ok := newPortNetworkPolicyRule(rule)
+		if !ok {
+			return PortNetworkPolicyRules{}, false
+		}
+		if len(newRule.L7Rules) > 0 {
+			rules.HaveL7Rules = true
+		}
+		if idx == 0 {
+			firstTypeName = typeName
+		} else if typeName != firstTypeName {
+			panic(fmt.Errorf("NPDS: Mismatching L7 types on the same port %v", config))
+		}
+		rules.Rules = rules.Rules[:idx+1] // Increase slice len
+		rules.Rules[idx] = newRule
+	}
+	return rules, true
+}
+
+func (p *PortNetworkPolicyRules) Matches(remoteId uint32, l7 interface{}) bool {
+	if !p.HaveL7Rules {
+		// If there are no L7 rules, host proxy will not create a proxy redirect at all,
+		// whereby the decicion made by the bpf datapath is final. Emulate the same behavior
+		// in the sidecar by allowing such traffic.
+		// TODO: This will need to be revised when non-bpf datapaths are to be supported.
+		log.Debugf("NPDS::PortNetworkPolicyRules: No L7 rules; matches (%v)", p)
+		return true
+	}
+	// Empty set matches any payload from anyone
+	if len(p.Rules) == 0 {
+		log.Debugf("NPDS::PortNetworkPolicyRules: No Rules; matches (%v)", p)
+		return true
+	}
+	for _, rule := range p.Rules {
+		if rule.Matches(remoteId, l7) {
+			log.Debugf("NPDS::PortNetworkPolicyRules(remoteId=%d): rule matches (%v)", remoteId, p)
+			return true
+		}
+	}
+	return false
+}
+
+type PortNetworkPolicies struct {
+	Rules map[uint32]PortNetworkPolicyRules
```

Comment:
From context I noticed this maps a port to portNetworkpolicyrules. Why isn't `uint32` a `uint16`? Also, what if in the future we allow rules per port/protocol pair, does this map still make sense?

Also, add comment to what this map maps into.

Reply:
Leave a comment that `uint32` is a port number that maps to `PortNetworkPolicyRules`